pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -834,20 +834,35 @@ func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1.Pipeline
 	recorder := controller.GetEventRecorder(ctx)
 
 	// nextRpts holds a list of pipeline tasks which should be executed next
-	nextRpts, err := pipelineRunFacts.DAGExecutionQueue()
+	// tmpNextRpts holds the nextRpts temporarily,
+	// tmpNextRpts is later filtered to check for the missing result reference
+	// if the pipelineTask is valid then it is added to the nextRpts
+	tmpNextRpts, err := pipelineRunFacts.DAGExecutionQueue()
 	if err != nil {
 		logger.Errorf("Error getting potential next tasks for valid pipelinerun %s: %v", pr.Name, err)
 		return controller.NewPermanentError(err)
 	}
 
-	// Check for Missing Result References
-	err = resources.CheckMissingResultReferences(pipelineRunFacts.State, nextRpts)
-	if err != nil {
-		logger.Infof("Failed to resolve task result reference for %q with error %v", pr.Name, err)
-		pr.Status.MarkFailed(v1.PipelineRunReasonInvalidTaskResultReference.String(), err.Error())
-		return controller.NewPermanentError(err)
+	var nextRpts resources.PipelineRunState
+	for _, nextRpt := range tmpNextRpts {
+		// Check for Missing Result References and
+		// store the faulty task in missingRefTask
+		missingRefTask, err := resources.CheckMissingResultReferences(pipelineRunFacts.State, nextRpt)
+		if err != nil {
+			logger.Infof("Failed to resolve task result reference for %q with error %v", pr.Name, err)
+			pr.Status.MarkFailed(v1.PipelineRunReasonInvalidTaskResultReference.String(), err.Error())
+			// check if pipeline contains finally tasks
+			// return the permanent error only if there is no finally task
+			fTaskNames := pipelineRunFacts.GetFinalTaskNames()
+			pipelineRunFacts.ValidationFailedTask = append(pipelineRunFacts.ValidationFailedTask, missingRefTask)
+			if len(fTaskNames) == 0 {
+				return controller.NewPermanentError(err)
+			}
```

Comment:
@afrittoli I got confused on a few things.
If any of the tasks is failed in the validation then do we need to empty the `nextRpts`?
If yes, then I am not getting why we should empty the `nextRpts` list if not final task is defined.

Reply:
You are right, sorry if my comment was confusing, if any task fails validation, then `nextRpts` should be emptied, regardless of whether there are final tasks or not.

The reason for setting it to empty is that there is logic later in the code that will process this list, so the content of the list should represent what is going to happen in terms of execution.

In case of failed validation, no new task (from the non-finally section of the pipeline) will be scheduled, hence the list should be empty.

We should make sure we have enough test coverage to verify that everything works correctly with this changes

Reply:
Yes, that makes sense.
Let me try with the changes.

Reply:
@afrittoli as per the above discussion I have removed the `NewPermanentError` but the unit test named `TestMissingResultWhenStepErrorIsIgnored` is failing with the error
> pipelinerun_test.go:1310: Expected the error to be permanent: true but got: false

So it is expecting the NewPermanentError but we are not throwing any error.  

Relevant test: https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun_test.go#L1164
Failing at: https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun_test.go#L8434

Can we check if the finally task exists then don't return the error else return the error (as reconciliation can stop here)?

Reply:
The unit test fails because it was designed to verify the old behaviour. 
If we change the logic we need to update the unit test accordingly.

The old logic was: 
- if we encounter a task with invalid result references, fail the pipeline. Nothing else happens

The implementation was like that because we had no way to express that the Task failing validation actually failed (because there is no TaskRun created), thus the existing logic to mark the pipeline as failed and terminate it would not work.

The new logic is:
- if we encounter a task with invalid result references, fail the task by adding it to a list of tasks failed by validation
- continue pipeline execution

The new logic also means that:
- because a Task is marked as failed, no new Task from the main pipeline will be scheduled
- if finally Tasks are available they will be scheduled
- if no finally Tasks are available, there's nothing else to do, the pipeline is marked as finished and failed (because of the task validation failure)

Reply:
Thanks for the detailed explanation. I understand what you are trying to say but I came across this while changing the unit tests and it says:
> // When a PipelineRun is invalid and can't run, we expect a permanent error that will
		// tell the Reconciler to not keep trying to reconcile.
		
Ref: https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun_test.go#L8428
So, I am not sure if we should mark it as a permanent error if finally is not present inside `runNextSchedulableTask` or if it should be handled in a separate part of the code.